librad/src/meta/entity.rs

```diff
@@ -381,69 +593,35 @@ where
     }
     /// Check certifier presence
     fn has_certifier(&self, c: &RadUrn) -> bool {
-        self.certifiers.contains(c)
+        self.certifiers.contains_key(c)
     }
 
     /// Turn the entity into its canonical data representation
-    /// (for hashing or signing)
+    /// (for hashing)
     pub fn canonical_data(&self) -> Result<Vec<u8>, Error> {
         self.to_data().canonical_data()
     }
 
-    /// Check that this key is allowed to sign the entity by checking that the
-    /// `by` argument is correct:
-    ///
-    /// - either the key is owned by `self`
-    /// - or it belongs to the given certifier
-    pub async fn check_key(
-        &self,
-        key: &PublicKey,
-        by: &Signatory,
-        resolver: &impl Resolver<User<Draft>>,
-    ) -> Result<(), Error> {
-        match by {
-            Signatory::OwnedKey => {
-                if !self.has_key(key) {
-                    return Err(Error::KeyNotPresent(key.to_owned()));
-                }
-            },
-            Signatory::User(uri) => {
-                let user = resolver.resolve(&uri).await?;
-                if !user.has_key(key) {
-                    return Err(Error::UserKeyNotPresent(uri.to_owned(), key.to_owned()));
-                }
-            },
-        }
-        Ok(())
-    }
-
-    /// Given a private key, compute the signature of the entity canonical data
-    /// FIXME[ENTITY]: we should check the hash instead: it is cheaper and makes
-    /// also verification way faster because we would not need to rebuild the
-    /// canonical data at every check (we can trust the hash correctness)
-    pub fn compute_signature(&self, key: &SecretKey) -> Result<Signature, Error> {
-        Ok(key.sign(&self.canonical_data()?))
-    }
-
     /// Given a private key owned by the entity, sign the current entity
     ///
     /// The following checks are performed:
     ///
     /// - the entity has not been already signed using this same key
     /// - this key is owned by the current entity
     pub fn sign_owned(&mut self, key: &SecretKey) -> Result<(), Error> {
+        let hash = self.hash().clone();
+        let revision = self.revision;
         let public_key = key.public();
-        if self.signatures().contains_key(&public_key) {
-            return Err(Error::SignatureAlreadyPresent(public_key));
-        }
-        if !self.has_key(&public_key) {
-            return Err(Error::KeyNotPresent(public_key));
-        }
-        let signature = EntitySignature {
-            by: Signatory::OwnedKey,
-            sig: self.compute_signature(key)?,
-        };
-        self.signatures.insert(public_key, signature);
+        self.keys
+            .get_mut(&public_key)
+            .ok_or_else(|| Error::KeyNotPresent(public_key.clone()))
+            .map(|opt| match opt {
+                Some(_) => Err(Error::SignatureAlreadyPresent(public_key)),
```

Comment:
In principle, yes, it should be a no-op.
I did it like this because the API user should never attempt to "re-sign" something, it smells of application error, or trying to change an already valid signature, which does not make sense.